### PR TITLE
Sneak attack 2024

### PR DESCRIPTION
**Pull request by dmportella**
Huge change!

Sorry

ermm sooo....

added new dialog double dialog
support for Devious Strikes
support for Improved Cunning Strike
support for Supreme Sneak
support for Cunning Strike
added action capture for features so we can get the additional information on top of the choices and options
added get class feature choices
added simple roll-info for cunning strike reusing save dc section to show selected effects
added setting auto turn off applies only if turned on
added cunning string infor to attack and spell rolls to support shadow blade
bunch of fixes for the sneak attack and also to support criticals
simplified querycunningstrike
added settings to change to build attack roll
added setting to popup
small fix when effects cost exceed the sneak die count
small fix where the cancel buttons is selected
added getsneakattackactions
simplified code base
fixed small bug on cancel button selection

# New setting

> New setting available it is off by default, cunning strike is only applied if the user wants to apply otherwise it will use the default sneak attack.

image

> settings can be hotkeyed

image

# New Cunning Strike Dialog

>New Dialog appears ONLY if the user turned on the settings above, it will be populated with the character cunning strike actions they have access too and when items are selected it will reduce the sneak attack die count accordingly.

image
image

> If the have improved cunning strike the dialog will present them with two combo boxes 

image
image

**THIS ALLOWS THE USER TO SELECT THE SAME VALUES**
**THE SNEAK DIE COUNT WILL BE DEDUCTED ACCORDINGLY**

> There is not rules in dnd 5d 2024 stating that the use can not select the same effect twice.

> when selecting it twice the same effect the vtt will show it

image

> selecting only one effect (whether you can select only one or more) will only show the effect you selected

image
image

> die count is reduced normally

image
image

> selecting more effect where the cost is higher then the sneak die count it will be ignored and the sneak will NOT be rolled 

image

# Shadow Blade

> shadow blade is supported

image

# Sneak attack action 

> added support for the sneak attack actions so if the cunning strike is turned one the action can be used with the new feature

image
image
image

# Action lists capture when updating features

> The code now takes into account action lists, a new css tag that works the same way as choices and options

image

# Homebrew Support

> added support to homebrew, as long that the homebrew class feature, racial trait or feat has an action that matches the regex format for the sneak attack cunning strike effect, it will be shown in the dialog box.

> `Sneak Attack: <name> (Cost: <X>D6)`

Screenshot 2024-11-23 233143

> this example feature that add a new effect called DEATH!!!

> this is how looks on the character sheet

Screenshot 2024-11-23 234220

> the action added to the class feature in example

Screenshot 2024-11-24 003406

> How it looks like in the edit window, needs right name and action set in activation type

Screenshot 2024-11-24 003352

> dialog with new sneak attack effect

Screenshot 2024-11-24 003224

> Vtt shows the new effect

Screenshot 2024-11-24 004250

### CODE REVIEW

**Review by kakaroto — COMMENTED** [review]
Amazing work, thank you so much!
I've had a few comments but only about coding style, I've fixed it all and pushed to master. I'd appreciate if you could review my commit here: https://github.com/kakaroto/Beyond20/commit/a354d6c107fd50117a2fce8c2dcc6df6ce10cdfd
Also, if you have any comments/disagree with any of my review comments, let me know.
(FYI: I admit to not having tested the changes I made, it's now 5AM and I gotta sleep)

File: src/dndbeyond/base/utils.js [hunk 1]
```diff
@@ -130,9 +130,43 @@ async function queryDamageTypeFromArray(name, damages, damage_types, possible_ty
...
+async function queryCunningStrike() {
+    const selection = [];
+    // Sneak Attack free-rules, pg. 129
+    // Cunning Strike free-rules, pg. 130
+    const actions = [{ action: "None" }];
+    const hasImprovedCunningStrike = character.hasClassFeature("Improved Cunning Strike");
+    
+    actions.push(...character.getSneakAttackActions());
+    
+    const options = [...actions.map(m => m["die"] ? `${m.action}: ${m.die}D6` : `${m.action}`)];
```
[the remark this answers is not in the document]
Comment (dmportella) [reply src/dndbeyond/base/utils.js:142]: i think i was getting double vision by that point XD

File: src/dndbeyond/base/utils.js [hunk 2]
```diff
@@ -226,6 +260,55 @@ async function buildAttackRoll(character, attack_source, name, description, prop
             const choice = await queryDamageTypeFromArray(roll_properties.name, damages, damage_types, ["Radiant", "Necrotic"]);
             if (choice === null) return null; // Query was cancelled;
         }
+
+        if (character.getSetting("rogue-sneak-attack", false) && character.hasClass("Rogue") && character.hasClassFeature("Sneak Attack 2024") && !name.includes("Psionic Power: Psychic Whispers") && 
+            (properties["Attack Type"] == "Ranged" ||
+            (properties["Properties"] && properties["Properties"].includes("Finesse")) ||
+            (name && (name.includes("Psychic Blade") || name.includes("Shadow Blade") || name.includes("Sneak Attack"))))) {
```
[the remark this answers is not in the document]
Comment (dmportella) [reply src/dndbeyond/base/utils.js:267]: yep probably not needed that is oke

File: src/dndbeyond/base/utils.js [hunk 3]
```diff
@@ -226,6 +260,55 @@ async function buildAttackRoll(character, attack_source, name, description, prop
...
+                const sneakDieCount = Math.ceil(character._classes["Rogue"] / 2);
+                const sneak_attack = sneakDieCount + "d6";
+                let selectionErrorDontApplySneak = false;
```
[the remark this answers is not in the document]
Comment (dmportella) [reply src/dndbeyond/base/utils.js:270]: that is a better option then not doing it

File: src/dndbeyond/content-scripts/character.js [hunk 4]
```diff
@@ -741,14 +741,15 @@ function handleSpecialWeaponAttacks(damages=[], damage_types=[], properties, set
...
-            (properties["Attack Type"] == "Ranged" ||
+        if(character.getSetting("rogue-sneak-attack", false) && (properties["Attack Type"] == "Ranged" ||
             (properties["Properties"] && properties["Properties"].includes("Finesse")) ||
-            (action_name && (action_name.includes("Psychic Blade") || action_name.includes("Shadow Blade"))))) {
-            const sneak_attack = Math.ceil(character._classes["Rogue"] / 2) + "d6";
-            damages.push(sneak_attack);
-            damage_types.push("Sneak Attack");
-        }
+            (action_name && (action_name.includes("Psychic Blade") || action_name.includes("Shadow Blade"))))){
+            if (character.hasClassFeature("Sneak Attack")) {
```
[the remark this answers is not in the document]
Comment (dmportella) [reply src/dndbeyond/content-scripts/character.js:747]: yep that is correct

**Comment by dmportella** [comment]
> Amazing work, thank you so much! I've had a few comments but only about coding style, I've fixed it all and pushed to master. I'd appreciate if you could review my commit here: [a354d6c](https://github.com/kakaroto/Beyond20/commit/a354d6c107fd50117a2fce8c2dcc6df6ce10cdfd) Also, if you have any comments/disagree with any of my review comments, let me know. (FYI: I admit to not having tested the changes I made, it's now 5AM and I gotta sleep)

no worries, i will get latest and test everything.
thank you for doing all the reviews etc
<3

File: src/dndbeyond/base/utils.js [hunk 5]
```diff
@@ -226,6 +260,55 @@ async function buildAttackRoll(character, attack_source, name, description, prop
             const choice = await queryDamageTypeFromArray(roll_properties.name, damages, damage_types, ["Radiant", "Necrotic"]);
             if (choice === null) return null; // Query was cancelled;
         }
+
+        if (character.getSetting("rogue-sneak-attack", false) && character.hasClass("Rogue") && character.hasClassFeature("Sneak Attack 2024") && !name.includes("Psionic Power: Psychic Whispers") &&
```
[the remark this answers is not in the document]
Comment (dmportella) [reply src/dndbeyond/base/utils.js:264]: moving it there will break the sneak attack action and shadow blade spell the reason to be on the build attack roll is that the sneak attack can appear in multiple places rather then just weapon attacks, we can refactor into a function and just place it on all areas tho

[the remark this answers is not in the document]
Comment (kakaroto) [reply src/dndbeyond/base/utils.js:264]: oh, I thought I deleted that comment. I had started moving it during the refactor and noticed the issues and understood why you put it in utils, so I moved it back and thought I deleted this review comment... I wonder if I deleted another one by mistake 🤔 maybe I just forgot to delete 😁

**Pull request merged**